go/vt/vttablet/tabletserver/query_executor_test.go

```diff
@@ -47,982 +48,366 @@ import (
-	db := setUpQueryExecutorTest(t)
-	defer db.Close()
-	query := "update test_table set addr = 3 where name = 1"
-	expandedQuery := "select pk from test_table where name = 1 limit 10001 for update"
-	want := &sqltypes.Result{}
-	db.AddQuery(query, want)
-	db.AddQuery(expandedQuery, &sqltypes.Result{
-		Fields: []*querypb.Field{
-			{Type: sqltypes.Int32},
-		},
-		RowsAffected: 1,
-		Rows: [][]sqltypes.Value{
-			{sqltypes.NewInt32(2)},
-		},
-	})
-	updateQuery := "update test_table set addr = 3 where pk in (2)"
-	db.AddQuery(updateQuery, want)
-	ctx := context.Background()
-	tsv := newTestTabletServer(ctx, noFlags, db)
-	txid := newTransaction(tsv, nil)
-	qre := newTestQueryExecutor(ctx, tsv, query, txid)
-	tsv.qe.binlogFormat = connpool.BinlogFormatRow
-	defer tsv.StopService()
-	defer testCommitHelper(t, tsv, qre)
-	checkPlanID(t, planbuilder.PlanDMLSubquery, qre.plan.PlanID)
-	got, err := qre.Execute()
-	if err != nil {
-		t.Fatalf("qre.Execute() = %v, want nil", err)
-	}
-	if !reflect.DeepEqual(got, want) {
-		t.Fatalf("got: %v, want: %v", got, want)
-	}
-	wantqueries := []string{updateQuery}
-	gotqueries := fetchRecordedQueries(qre)
-	if !reflect.DeepEqual(gotqueries, wantqueries) {
-		t.Errorf("queries: %v, want %v", gotqueries, wantqueries)
+func TestQueryExecutorDMLLimitFailure(t *testing.T) {
+	type dbResponse struct {
+		query  string
+		result *sqltypes.Result
+	}
+
+	dmlResult := &sqltypes.Result{
+		RowsAffected: 3,
+	}
+
+	// The queries are run both in and outside a transaction.
+	testcases := []struct {
+		input       string
+		passThrough bool
+		dbResponses []dbResponse
+		planWant    string
+		logWant     string
+		inTxWant    string
+	}{{
+		input: "update test_table set a=1",
+		dbResponses: []dbResponse{{
+			query:  "update test_table set a = 1 limit 3",
+			result: dmlResult,
+		}},
+		logWant:  "begin; update test_table set a = 1 limit 3; rollback",
+		inTxWant: "update test_table set a = 1 limit 3; rollback",
+	}, {
+		input: "delete from test_table",
+		dbResponses: []dbResponse{{
+			query:  "delete from test_table limit 3",
+			result: dmlResult,
+		}},
+		logWant:  "begin; delete from test_table limit 3; rollback",
+		inTxWant: "delete from test_table limit 3; rollback",
+	}}
+	for _, tcase := range testcases {
+		func() {
+			db := setUpQueryExecutorTest(t)
+			defer db.Close()
+			for _, dbr := range tcase.dbResponses {
+				db.AddQuery(dbr.query, dbr.result)
+			}
+			ctx := callerid.NewContext(context.Background(), callerid.NewEffectiveCallerID("a", "b", "c"), callerid.NewImmediateCallerID("d"))
+			tsv := newTestTabletServer(ctx, smallResultSize, db)
+			defer tsv.StopService()
+
+			tsv.SetPassthroughDMLs(tcase.passThrough)
+
+			// Test outside a transaction.
+			qre := newTestQueryExecutor(ctx, tsv, tcase.input, 0)
+			_, err := qre.Execute()
+			wantErr := "caller id: d: row count exceeded 2 (errno 10001) (sqlstate HY000)"
+			if err == nil || err.Error() != wantErr {
```

Comment:
nit: You could use
```
require.Error(t, wantErr)
```

Reply:
For some reason, assert didn't work. In any case, I had to change the test to use "Contains".
But `assert.Contains` can't check if an error contains a string, which is unfortunate.